src/controllers/engine/controllerengine.h

```diff
@@ -167,21 +155,31 @@ class ControllerEngine : public QObject {
     void errorDialogButton(const QString& key, QMessageBox::StandardButton button);
 
   private:
-    bool syntaxIsValid(const QString& scriptCode);
-    bool evaluate(const QString& scriptName, QList<QString> scriptPaths);
-    bool internalExecute(QScriptValue thisObject, const QString& scriptCode);
-    bool internalExecute(QScriptValue thisObject, QScriptValue functionObject,
-                         QScriptValueList arguments);
+    bool evaluateScriptFile(const QString& scriptName, QList<QString> scriptPaths);
+    bool internalExecute(QJSValue thisObject, const QString& scriptCode);
+    bool internalExecute(const QString& scriptCode);
+    bool internalExecute(QJSValue thisObject, QJSValue functionObject,
+                         QJSValueList arguments);
+    bool internalExecute(QJSValue functionObject,
+                         QJSValueList arguments);
     void initializeScriptEngine();
 
     void scriptErrorDialog(const QString& detailedError);
     void generateScriptFunctions(const QString& code);
     // Stops and removes all timers (for shutdown).
     void stopAllTimers();
 
-    void callFunctionOnObjects(QList<QString>, const QString&, QScriptValueList args = QScriptValueList());
-    bool checkException();
-    QScriptEngine *m_pEngine;
+    void callFunctionOnObjects(QList<QString>, const QString&, QJSValueList args = QJSValueList());
+    // Convert a byteArray to a JS typed array over an ArrayBuffer
+    QJSValue byteArrayToScriptValue(const QByteArray byteArray);
+    // Throws EvaluationException and NullEngineException.
+    QJSValue evaluateProgram(const QString& program, const QString& fileName = QString(),
+    		int lineNumber = 1);
+
+    // Shows a UI dialog notifying of an script evaluation error.
+    // Precondition: QJSValue.isError() == true
```

Comment:
Yes, the precondition is checked on every place this method is called.

Reply:
I mean check it in this method and return an error/do something sensible if it's not met.

Reply:
This would mean adding inside the function the following check:
```
if (!evaluationResult.isError()) {
    return;
}
```
We can't remove the checks that surround each call to `showScriptExceptionDialog(...)`, because the calling code returns with a different value. (see controllerengine.cpp:929)
We would then be checking this twice, one before before each method call and one inside the method.

Reply:
I'm just trying to prevent future bugs. What happens if this function is called when `QJSValue.isError() != true`?

Reply:
Undefined behavior :)

I guess we can assert the value is an error inside a function. Thanks for the feedback.

Reply:
No problem. Undefined behavior has bitten us many times in the past. Our future selves thank you. :wink: